File: src/main/java/me/pcasaes/hexoids/configuration/HexoidConfigurations.java
package me.pcasaes.hexoids.configuration;


import io.quarkus.runtime.StartupEvent;
import me.pcasaes.hexoids.core.domain.config.Config;
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndex;
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndexFactory;
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndex;
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndexFactory;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import javax.annotation.PostConstruct;
import javax.annotation.Priority;
import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Observes;
import javax.inject.Inject;
import javax.interceptor.Interceptor;
import java.util.logging.Logger;

@ApplicationScoped
public class HexoidConfigurations {

    private static final Logger LOGGER = Logger.getLogger(HexoidConfigurations.class.getName());

    private float inertiaDampenCoefficient;

    private long updateFrequencyInMillis;

    private float minMove;

    private float playerMaxMove;

    private float playerMaxAngleDivisor;

    private long expungeSinceLastSpawnTimeout;

    private String playerResetPosition;

    private int maxBolts;

    private int playerNameLength;

    private int boltMaxDuration;

    private float boltSpeed;

    private float boltCollisionRadius;

    private boolean boltInertiaEnabled;

    private float boltInertiaRejectionScale;

    private float boltInertiaProjectionScale;

    private float boltInertiaNegativeProjectionScale;

    private float playerDestroyedShockwaveDistance;

    private long playerDestroyedShockwaveDuration;

    private float playerDestroyedShockwaveImpulse;


    /**
     * This singleton will eager load before all others
     *
     * @param event
     */
    public void startup(@Observes @Priority(Interceptor.Priority.APPLICATION) StartupEvent event) {
        LOGGER.info("Eager load Configuration");
    }

    @PostConstruct
    public void start() {
        LOGGER.info("hexoids.config.inertia.dampen-coefficients=" + getInertiaDampenCoefficient());
        LOGGER.info("hexoids.config.update-frequency-in-millis=" + getUpdateFrequencyInMillis());
        LOGGER.info("hexoids.config.min.min=" + getMinMove());
        LOGGER.info("hexoids.config.player.expunge-since-last-spawn-timeout=" + getExpungeSinceLastSpawnTimeout());
        LOGGER.info("hexoids.config.player.name-length=" + getPlayerNameLength());
        LOGGER.info("hexoids.config.player.max.move=" + getPlayerMaxMove());
        LOGGER.info("hexoids.config.player.max.bolts=" + getMaxBolts());
        LOGGER.info("hexoids.config.player.max.angle.divisors=" + getPlayerMaxAngleDivisor());
        LOGGER.info("hexoids.config.player.reset.position=" + getPlayerResetPosition());
        LOGGER.info("hexoids.config.player.destroyed.shockwave.distance=" + getPlayerDestroyedShockwaveDistance());
        LOGGER.info("hexoids.config.player.destroyed.shockwave.duration.ms=" + getPlayerDestroyedShockwaveDuration());
        LOGGER.info("hexoids.config.player.destroyed.shockwave.impulse=" + getPlayerDestroyedShockwaveImpulse());
        LOGGER.info("hexoids.config.bolt.max.duration=" + getBoltMaxDuration());
        LOGGER.info("hexoids.config.bolt.speed=" + getBoltSpeed());
        LOGGER.info("hexoids.config.bolt.collision.radius=" + getBoltCollisionRadius());
        LOGGER.info("hexoids.config.bolt.inertia.enabled=" + isBoltInertiaEnabled());
        LOGGER.info("hexoids.config.bolt.inertia.rejection-scale=" + getBoltInertiaRejectionScale());
        LOGGER.info("hexoids.config.bolt.inertia.projection-scale=" + getBoltInertiaProjectionScale());
        LOGGER.info("hexoids.config.bolt.inertia.negative-projection-scale=" + getBoltInertiaNegativeProjectionScale());

        Config.get().setUpdateFrequencyInMillis(getUpdateFrequencyInMillis());
        Config.get().setInertiaDampenCoefficient(getInertiaDampenCoefficient());
        Config.get().setMaxBolts(getMaxBolts());
        Config.get().setMinMove(getMinMove());
        Config.get().setExpungeSinceLastSpawnTimeout(getExpungeSinceLastSpawnTimeout());
        Config.get().setPlayerNameLength(getPlayerNameLength());
        Config.get().setPlayerMaxMove(getPlayerMaxMove());
        Config.get().setPlayerMaxAngleDivisor(getPlayerMaxAngleDivisor());
        Config.get().setPlayerResetPosition(getPlayerResetPosition());
        Config.get().setBoltMaxDuration(getBoltMaxDuration());
        Config.get().setBoltSpeed(getBoltSpeed());
        Config.get().setBoltCollisionRadius(getBoltCollisionRadius());
        Config.get().setBoltInertiaEnabled(isBoltInertiaEnabled());
        Config.get().setBoltInertiaRejectionScale(getBoltInertiaRejectionScale());
        Config.get().setBoltInertiaProjectionScale(getBoltInertiaProjectionScale());
        Config.get().setBoltInertiaNegativeProjectionScale(getBoltInertiaNegativeProjectionScale());
        Config.get().getPlayerDestroyedShockwave().setDistance(getPlayerDestroyedShockwaveDistance());
        Config.get().getPlayerDestroyedShockwave().setDuration(getPlayerDestroyedShockwaveDuration());
        Config.get().getPlayerDestroyedShockwave().setImpulse(getPlayerDestroyedShockwaveImpulse());
    }


    public float getInertiaDampenCoefficient() {
        return inertiaDampenCoefficient;
    }

    @Inject
    public void setPlayerSpatialIndex(PlayerSpatialIndex playerSpatialIndex) {
        PlayerSpatialIndexFactory.factory().setPlayerSpatialIndex(playerSpatialIndex);
    }

    @Inject
    public void setBarrierSpatialIndex(BarrierSpatialIndex barrierSpatialIndex) {
        BarrierSpatialIndexFactory.factory().setBarrierSpatialIndex(barrierSpatialIndex);
    }

    @Inject
    public void setInertiaDampenCoefficient(
            @ConfigProperty(
                    name = "hexoids.config.inertia.dampen-coefficient",
                    defaultValue = "-0.001"
            ) float inertiaDampenCoefficient) {
        this.inertiaDampenCoefficient = inertiaDampenCoefficient;
    }

    public long getUpdateFrequencyInMillis() {
        return updateFrequencyInMillis;
    }

    @Inject
    public void setUpdateFrequencyInMillis(
            @ConfigProperty(
                    name = "hexoids.config.update-frequency-in-millis",
                    defaultValue = "50"
            ) long updateFrequencyInMillis) {
        this.updateFrequencyInMillis = updateFrequencyInMillis;
    }

    public float getMinMove() {
        return minMove;
    }

    @Inject
    public void setMinMove(
            @ConfigProperty(
                    name = "hexoids.config.min.move",
                    defaultValue = "0.0001"
            ) float minMove) {
        this.minMove = minMove;
    }

    public float getPlayerMaxMove() {
        return playerMaxMove;
    }

    @Inject
    public void setPlayerMaxMove(
            @ConfigProperty(
                    name = "hexoids.config.player.max.move",
                    defaultValue = "10"
            ) float playerMaxMove) {
        this.playerMaxMove = playerMaxMove;
    }

    public float getPlayerMaxAngleDivisor() {
        return playerMaxAngleDivisor;
    }

    @Inject
    public void setPlayerMaxAngleDivisor(
            @ConfigProperty(
                    name = "hexoids.config.player.max.angle.divisor",
                    defaultValue = "4"
            ) float playerMaxAngleDivisor) {
        this.playerMaxAngleDivisor = playerMaxAngleDivisor;
    }

    public long getExpungeSinceLastSpawnTimeout() {
        return expungeSinceLastSpawnTimeout;
    }

    @Inject

    public void setExpungeSinceLastSpawnTimeout(
            @ConfigProperty(
                    name = "hexoids.config.player.expunge-since-last-spawn-timeout",
                    defaultValue = "60000"
            ) long expungeSinceLastSpawnTimeout) {
        this.expungeSinceLastSpawnTimeout = expungeSinceLastSpawnTimeout;
    }

    public String getPlayerResetPosition() {
        return playerResetPosition;
    }

    @Inject
    public void setPlayerResetPosition(
            @ConfigProperty(
                    name = "hexoids.config.player.reset.position",
                    defaultValue = "rng"
            ) String playerResetPosition) {
        this.playerResetPosition = playerResetPosition;
    }

    public int getMaxBolts() {
        return maxBolts;
    }

    @Inject
    public void setMaxBolts(
            @ConfigProperty(
                    name = "hexoids.config.player.max.bolts",
                    defaultValue = "10"
            ) int maxBolts) {
        this.maxBolts = maxBolts;
    }

    public int getPlayerNameLength() {
        return playerNameLength;
    }

    @Inject
    public void setPlayerNameLength(
            @ConfigProperty(
                    name = "hexoids.config.player.name-length",
                    defaultValue = "7"
            ) int playerNameLength) {
        this.playerNameLength = playerNameLength;
    }

    public int getBoltMaxDuration() {
        return boltMaxDuration;
    }

    @Inject
    public void setBoltMaxDuration(
            @ConfigProperty(
                    name = "hexoids.config.bolt.max.duration",
                    defaultValue = "10000"
            ) int boltMaxDuration) {
        this.boltMaxDuration = boltMaxDuration;
    }

    public float getBoltSpeed() {
        return boltSpeed;
    }

    @Inject
    public void setBoltSpeed(
            @ConfigProperty(
                    name = "hexoids.config.bolt.speed",
                    defaultValue = "0.07"
            ) float boltSpeed) {
        this.boltSpeed = boltSpeed;
    }

    public float getBoltCollisionRadius() {
        return boltCollisionRadius;
    }

    @Inject
    public void setBoltCollisionRadius(
            @ConfigProperty(
                    name = "hexoids.config.bolt.collision.radius",
                    defaultValue = "0.001"
            ) float boltCollisionRadius) {
        this.boltCollisionRadius = boltCollisionRadius;
    }

    public boolean isBoltInertiaEnabled() {
        return boltInertiaEnabled;
    }

    @Inject
    public void setBoltInertiaEnabled(
            @ConfigProperty(
                    name = "hexoids.config.bolt.inertia.enabled",
                    defaultValue = "true"
            ) boolean boltInertiaEnabled) {
        this.boltInertiaEnabled = boltInertiaEnabled;
    }

    public float getBoltInertiaRejectionScale() {
        return boltInertiaRejectionScale;
    }

    @Inject
    public void setBoltInertiaRejectionScale(
            @ConfigProperty(
                    name = "hexoids.config.bolt.inertia.rejection-scale",
                    defaultValue = "0.8"
            ) float boltInertiaRejectionScale) {
        this.boltInertiaRejectionScale = boltInertiaRejectionScale;
    }

    public float getBoltInertiaProjectionScale() {
        return boltInertiaProjectionScale;
    }

    @Inject
    public void setBoltInertiaProjectionScale(
            @ConfigProperty(
                    name = "hexoids.config.bolt.inertia.projection-scale",
                    defaultValue = "0.8"
            ) float boltInertiaProjectionScale) {
        this.boltInertiaProjectionScale = boltInertiaProjectionScale;
    }

    public float getBoltInertiaNegativeProjectionScale() {
        return boltInertiaNegativeProjectionScale;
    }

    @Inject
    public void setBoltInertiaNegativeProjectionScale(
            @ConfigProperty(
                    name = "hexoids.config.bolt.inertia.negative-projection-scale",
                    defaultValue = "0.1"
            ) float boltInertiaNegativeProjectionScale) {
        this.boltInertiaNegativeProjectionScale = boltInertiaNegativeProjectionScale;
    }

    public float getPlayerDestroyedShockwaveDistance() {
        return playerDestroyedShockwaveDistance;
    }

    @Inject
    public void setPlayerDestroyedShockwaveDistance(
            @ConfigProperty(
                    name = "hexoids.config.player.destroyed.shockwave.distance",
                    defaultValue = "0.0408"
            ) float playerDestroyedShockwaveDistance) {
        this.playerDestroyedShockwaveDistance = playerDestroyedShockwaveDistance;
    }

    public long getPlayerDestroyedShockwaveDuration() {
        return playerDestroyedShockwaveDuration;
    }

    @Inject
    public void setPlayerDestroyedShockwaveDuration(
            @ConfigProperty(
                    name = "hexoids.config.player.destroyed.shockwave.duration.ms",
                    defaultValue = "400"
            ) long playerDestroyedShockwaveDuration) {
        this.playerDestroyedShockwaveDuration = playerDestroyedShockwaveDuration;
    }

    public float getPlayerDestroyedShockwaveImpulse() {
        return playerDestroyedShockwaveImpulse;
    }

    @Inject
    public void setPlayerDestroyedShockwaveImpulse(@ConfigProperty(
            name = "hexoids.config.player.destroyed.shockwave.impulse",
            defaultValue = "0.007"
    ) float playerDestroyedShockwaveImpulse) {
        this.playerDestroyedShockwaveImpulse = playerDestroyedShockwaveImpulse;
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/configuration/HexoidConfigurations.kt
package me.pcasaes.hexoids.configuration

import io.quarkus.runtime.StartupEvent
import jakarta.annotation.PostConstruct
import jakarta.annotation.Priority
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.event.Observes
import jakarta.inject.Inject
import jakarta.interceptor.Interceptor
import me.pcasaes.hexoids.core.domain.config.Config
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndex
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndexFactory
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndex
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndexFactory
import org.eclipse.microprofile.config.inject.ConfigProperty
import java.beans.IntrospectionException
import java.beans.Introspector
import java.util.logging.Level
import java.util.logging.Logger


@ApplicationScoped
class HexoidConfigurations {
    private var inertiaDampenCoefficient = 0f

    private var updateFrequencyInMillis: Long = 0

    private var minMove = 0f

    private var playerMaxMove = 0f

    private var playerMaxAngleDivisor = 0f

    private var expungeSinceLastSpawnTimeout: Long = 0

    private var playerResetPosition: String = ""

    private var maxBolts = 0

    private var playerNameLength = 0

    private var boltMaxDuration = 0

    private var boltSpeed = 0f

    private var boltCollisionRadius = 0f

    private var boltInertiaEnabled: Boolean = false

    private var boltInertiaRejectionScale = 0f

    private var boltInertiaProjectionScale = 0f

    private var boltInertiaNegativeProjectionScale = 0f

    private var playerDestroyedShockwaveDistance = 0f

    private var playerDestroyedShockwaveDuration: Long = 0

    private var playerDestroyedShockwaveImpulse = 0f

    private var blackholeEventHorizonRadius = 0f

    private var blackholeGravityRadius = 0f

    private var blackholeGravityImpulse = 0f

    private var blackholeDampenFactor = 0f

    private var blackholeGenesisProbabilityFactor = 0

    private var blackholeTeleportProbability = 0f


    /**
     * This singleton will eager load before all others
     *
     * @param event
     */
    fun startup(@Observes @Priority(Interceptor.Priority.APPLICATION) event: StartupEvent?) {
        LOGGER.info("Eager load Configuration")
    }

    @PostConstruct
    fun start() {
        if (LOGGER.isLoggable(Level.INFO)) {
            try {
                for (pd in Introspector.getBeanInfo(HexoidConfigurations::class.java).getPropertyDescriptors()) {
                    if (pd.getReadMethod() != null && pd.getWriteMethod() != null && ("class" != pd.getName())) {
                        val annotations = pd.getWriteMethod().parameterAnnotations
                        if (annotations.size == 1 && annotations[0].size == 1 &&
                            annotations[0][0] is ConfigProperty
                        ) {
                            val configProperty =
                                annotations[0][0] as ConfigProperty
                            LOGGER.info(configProperty.name + "=" + pd.getReadMethod().invoke(this))
                        }
                    }
                }
            } catch (ex: ReflectiveOperationException) {
                LOGGER.warning { "Could not introspect config. " + ex }
            } catch (ex: IntrospectionException) {
                LOGGER.warning { "Could not introspect config. " + ex }
            } catch (ex: RuntimeException) {
                LOGGER.warning { "Could not introspect config. " + ex }
            }
        }

        Config.setUpdateFrequencyInMillis(getUpdateFrequencyInMillis())
        Config.setInertiaDampenCoefficient(getInertiaDampenCoefficient())
        Config.setMaxBolts(getMaxBolts())
        Config.setMinMove(getMinMove())
        Config.setExpungeSinceLastSpawnTimeout(getExpungeSinceLastSpawnTimeout())
        Config.setPlayerNameLength(getPlayerNameLength())
        Config.setPlayerMaxMove(getPlayerMaxMove())
        Config.setPlayerMaxAngleDivisor(getPlayerMaxAngleDivisor())
        Config.setPlayerResetPosition(getPlayerResetPosition())
        Config.setBoltMaxDuration(getBoltMaxDuration())
        Config.setBoltSpeed(getBoltSpeed())
        Config.setBoltCollisionRadius(getBoltCollisionRadius())
        Config.setBoltInertiaEnabled(isBoltInertiaEnabled())
        Config.setBoltInertiaRejectionScale(getBoltInertiaRejectionScale())
        Config.setBoltInertiaProjectionScale(getBoltInertiaProjectionScale())
        Config.setBoltInertiaNegativeProjectionScale(getBoltInertiaNegativeProjectionScale())
        Config.getPlayerDestroyedShockwave().setDistance(getPlayerDestroyedShockwaveDistance())
        Config.getPlayerDestroyedShockwave().setDuration(getPlayerDestroyedShockwaveDuration())
        Config.getPlayerDestroyedShockwave().setImpulse(getPlayerDestroyedShockwaveImpulse())
        Config.getBlackhole().setDampenFactor(getBlackholeDampenFactor())
        Config.getBlackhole().setEventHorizonRadius(getBlackholeEventHorizonRadius())
        Config.getBlackhole().setGenesisProbabilityFactor(getBlackholeGenesisProbabilityFactor())
        Config.getBlackhole().setGravityImpulse(getBlackholeGravityImpulse())
        Config.getBlackhole().setGravityRadius(getBlackholeGravityRadius())
        Config.getBlackhole().setTeleportProbability(getBlackholeTeleportProbability())
    }


    fun getInertiaDampenCoefficient(): Float {
        return inertiaDampenCoefficient
    }

    @Inject
    fun setPlayerSpatialIndex(playerSpatialIndex: PlayerSpatialIndex) {
        PlayerSpatialIndexFactory.factory().setPlayerSpatialIndex(playerSpatialIndex)
    }

    @Inject
    fun setBarrierSpatialIndex(barrierSpatialIndex: BarrierSpatialIndex) {
        BarrierSpatialIndexFactory.factory().setBarrierSpatialIndex(barrierSpatialIndex)
    }

    @Inject
    fun setInertiaDampenCoefficient(
        @ConfigProperty(
            name = "hexoids.config.inertia.dampen-coefficient",
            defaultValue = "-0.001"
        ) inertiaDampenCoefficient: Float
    ) {
        this.inertiaDampenCoefficient = inertiaDampenCoefficient
    }

    fun getUpdateFrequencyInMillis(): Long {
        return updateFrequencyInMillis
    }

    @Inject
    fun setUpdateFrequencyInMillis(
        @ConfigProperty(
            name = "hexoids.config.update-frequency-in-millis",
            defaultValue = "50"
        ) updateFrequencyInMillis: Long
    ) {
        this.updateFrequencyInMillis = updateFrequencyInMillis
    }

    fun getMinMove(): Float {
        return minMove
    }

    @Inject
    fun setMinMove(
        @ConfigProperty(name = "hexoids.config.min.move", defaultValue = "0.0001") minMove: Float
    ) {
        this.minMove = minMove
    }

    fun getPlayerMaxMove(): Float {
        return playerMaxMove
    }

    @Inject
    fun setPlayerMaxMove(
        @ConfigProperty(name = "hexoids.config.player.max.move", defaultValue = "10") playerMaxMove: Float
    ) {
        this.playerMaxMove = playerMaxMove
    }

    fun getPlayerMaxAngleDivisor(): Float {
        return playerMaxAngleDivisor
    }

    @Inject
    fun setPlayerMaxAngleDivisor(
        @ConfigProperty(
            name = "hexoids.config.player.max.angle.divisor",
            defaultValue = "4"
        ) playerMaxAngleDivisor: Float
    ) {
        this.playerMaxAngleDivisor = playerMaxAngleDivisor
    }

    fun getExpungeSinceLastSpawnTimeout(): Long {
        return expungeSinceLastSpawnTimeout
    }

    @Inject
    fun setExpungeSinceLastSpawnTimeout(
        @ConfigProperty(
            name = "hexoids.config.player.expunge-since-last-spawn-timeout",
            defaultValue = "60000"
        ) expungeSinceLastSpawnTimeout: Long
    ) {
        this.expungeSinceLastSpawnTimeout = expungeSinceLastSpawnTimeout
    }

    fun getPlayerResetPosition(): String {
        return playerResetPosition
    }

    @Inject
    fun setPlayerResetPosition(
        @ConfigProperty(
            name = "hexoids.config.player.reset.position",
            defaultValue = "rng"
        ) playerResetPosition: String
    ) {
        this.playerResetPosition = playerResetPosition
    }

    fun getMaxBolts(): Int {
        return maxBolts
    }

    @Inject
    fun setMaxBolts(
        @ConfigProperty(name = "hexoids.config.player.max.bolts", defaultValue = "10") maxBolts: Int
    ) {
        this.maxBolts = maxBolts
    }

    fun getPlayerNameLength(): Int {
        return playerNameLength
    }

    @Inject
    fun setPlayerNameLength(
        @ConfigProperty(name = "hexoids.config.player.name-length", defaultValue = "7") playerNameLength: Int
    ) {
        this.playerNameLength = playerNameLength
    }

    fun getBoltMaxDuration(): Int {
        return boltMaxDuration
    }

    @Inject
    fun setBoltMaxDuration(
        @ConfigProperty(name = "hexoids.config.bolt.max.duration", defaultValue = "10000") boltMaxDuration: Int
    ) {
        this.boltMaxDuration = boltMaxDuration
    }

    fun getBoltSpeed(): Float {
        return boltSpeed
    }

    @Inject
    fun setBoltSpeed(
        @ConfigProperty(name = "hexoids.config.bolt.speed", defaultValue = "0.07") boltSpeed: Float
    ) {
        this.boltSpeed = boltSpeed
    }

    fun getBoltCollisionRadius(): Float {
        return boltCollisionRadius
    }

    @Inject
    fun setBoltCollisionRadius(
        @ConfigProperty(
            name = "hexoids.config.bolt.collision.radius",
            defaultValue = "0.001"
        ) boltCollisionRadius: Float
    ) {
        this.boltCollisionRadius = boltCollisionRadius
    }

    fun isBoltInertiaEnabled(): Boolean {
        return boltInertiaEnabled
    }

    @Inject
    fun setBoltInertiaEnabled(
        @ConfigProperty(name = "hexoids.config.bolt.inertia.enabled", defaultValue = "true") boltInertiaEnabled: Boolean
    ) {
        this.boltInertiaEnabled = boltInertiaEnabled
    }

    fun getBoltInertiaRejectionScale(): Float {
        return boltInertiaRejectionScale
    }

    @Inject
    fun setBoltInertiaRejectionScale(
        @ConfigProperty(
            name = "hexoids.config.bolt.inertia.rejection-scale",
            defaultValue = "0.8"
        ) boltInertiaRejectionScale: Float
    ) {
        this.boltInertiaRejectionScale = boltInertiaRejectionScale
    }

    fun getBoltInertiaProjectionScale(): Float {
        return boltInertiaProjectionScale
    }

    @Inject
    fun setBoltInertiaProjectionScale(
        @ConfigProperty(
            name = "hexoids.config.bolt.inertia.projection-scale",
            defaultValue = "0.8"
        ) boltInertiaProjectionScale: Float
    ) {
        this.boltInertiaProjectionScale = boltInertiaProjectionScale
    }

    fun getBoltInertiaNegativeProjectionScale(): Float {
        return boltInertiaNegativeProjectionScale
    }

    @Inject
    fun setBoltInertiaNegativeProjectionScale(
        @ConfigProperty(
            name = "hexoids.config.bolt.inertia.negative-projection-scale",
            defaultValue = "0.1"
        ) boltInertiaNegativeProjectionScale: Float
    ) {
        this.boltInertiaNegativeProjectionScale = boltInertiaNegativeProjectionScale
    }

    fun getPlayerDestroyedShockwaveDistance(): Float {
        return playerDestroyedShockwaveDistance
    }

    @Inject
    fun setPlayerDestroyedShockwaveDistance(
        @ConfigProperty(
            name = "hexoids.config.player.destroyed.shockwave.distance",
            defaultValue = "0.0408"
        ) playerDestroyedShockwaveDistance: Float
    ) {
        this.playerDestroyedShockwaveDistance = playerDestroyedShockwaveDistance
    }

    fun getPlayerDestroyedShockwaveDuration(): Long {
        return playerDestroyedShockwaveDuration
    }

    @Inject
    fun setPlayerDestroyedShockwaveDuration(
        @ConfigProperty(
            name = "hexoids.config.player.destroyed.shockwave.duration.ms",
            defaultValue = "400"
        ) playerDestroyedShockwaveDuration: Long
    ) {
        this.playerDestroyedShockwaveDuration = playerDestroyedShockwaveDuration
    }

    fun getPlayerDestroyedShockwaveImpulse(): Float {
        return playerDestroyedShockwaveImpulse
    }

    @Inject
    fun setPlayerDestroyedShockwaveImpulse(
        @ConfigProperty(
            name = "hexoids.config.player.destroyed.shockwave.impulse",
            defaultValue = "0.007"
        ) playerDestroyedShockwaveImpulse: Float
    ) {
        this.playerDestroyedShockwaveImpulse = playerDestroyedShockwaveImpulse
    }

    fun getBlackholeEventHorizonRadius(): Float {
        return blackholeEventHorizonRadius
    }

    @Inject
    fun setBlackholeEventHorizonRadius(
        @ConfigProperty(
            name = "hexoids.config.blackhole.eventhorizon.radius",
            defaultValue = "0.005"
        ) blackholeEventHorizonRadius: Float
    ) {
        this.blackholeEventHorizonRadius = blackholeEventHorizonRadius
    }

    fun getBlackholeGravityRadius(): Float {
        return blackholeGravityRadius
    }

    @Inject
    fun setBlackholeGravityRadius(
        @ConfigProperty(
            name = "hexoids.config.blackhole.gravity.radius",
            defaultValue = "0.07"
        ) blackholeGravityRadius: Float
    ) {
        this.blackholeGravityRadius = blackholeGravityRadius
    }

    fun getBlackholeGravityImpulse(): Float {
        return blackholeGravityImpulse
    }

    @Inject
    fun setBlackholeGravityImpulse(
        @ConfigProperty(
            name = "hexoids.config.blackhole.gravity.impulse",
            defaultValue = "0.07"
        ) blackholeGravityImpulse: Float
    ) {
        this.blackholeGravityImpulse = blackholeGravityImpulse
    }

    fun getBlackholeDampenFactor(): Float {
        return blackholeDampenFactor
    }

    @Inject
    fun setBlackholeDampenFactor(
        @ConfigProperty(
            name = "hexoids.config.blackhole.dampen.factor",
            defaultValue = "5"
        ) blackholeDampenFactor: Float
    ) {
        this.blackholeDampenFactor = blackholeDampenFactor
    }

    fun getBlackholeGenesisProbabilityFactor(): Int {
        return blackholeGenesisProbabilityFactor
    }

    @Inject
    fun setBlackholeGenesisProbabilityFactor(
        @ConfigProperty(
            name = "hexoids.config.blackhole.genesis.probability.factor",
            defaultValue = "3"
        ) blackholeGenesisProbabilityFactor: Int
    ) {
        this.blackholeGenesisProbabilityFactor = blackholeGenesisProbabilityFactor
    }

    fun getBlackholeTeleportProbability(): Float {
        return blackholeTeleportProbability
    }

    @Inject
    fun setBlackholeTeleportProbability(
        @ConfigProperty(
            name = "hexoids.config.blackhole.teleport.probability",
            defaultValue = "0.05"
        ) blackholeTeleportProbability: Float
    ) {
        this.blackholeTeleportProbability = blackholeTeleportProbability
    }

    companion object {
        private val LOGGER: Logger = Logger.getLogger(HexoidConfigurations::class.java.getName())
    }
}
